Add steps to createOffer and explicitly specify what is queued

adam-be opened the pull request:
Fixes #744 

This PR is currently based on PR #721 to make use of some changes to the generic queued operation section.

adam-be commented:
Rebased passed PR #721 so this PR should be easier to review now. @jan-ivar, can you please have a look?

jan-ivar commented on webrtc.html:
```diff
@@ -1251,30 +1251,99 @@
+                                produce an identity assertion, reject
+                                <var>p</var> with a <code>DOMException</code>
+                                object whose <code>name</code> attribute has
+                                the value <code>NotReadableError</code> and
```
comma

jan-ivar commented on webrtc.html:
```diff
@@ -1251,30 +1251,99 @@
+                                <p>Reject <var>p</var> with a
+                                <code>DOMException</code> object whose
+                                <code>name</code> attribute has the value
+                                <code>OperationError</code> and abort these
```
comma

jan-ivar commented on webrtc.html:
```diff
@@ -1251,30 +1251,99 @@
+                          <li>
+                            <p>If the process to generate an SDP offer failed
+                            for any reason, the User Agent MUST queue a task
+                            that runs the following steps:</p>
```
JSEP doesn't cover identity AFAIK, so logically does this need to say "If the process to generate an SDP offer failed, or if the identity provider was unable to produce an identity assertion, then ..." ?

jan-ivar commented:
Lgtm once `isClosed` check is moved out.

jan-ivar commented on webrtc.html:
```diff
@@ -1251,30 +1251,99 @@
+                      <li>
+                        <p>If <var>connection</var>'s [[<a>isClosed</a>]] slot
+                        is <code>true</code>, return a promise rejected with an
+                        <code>InvalidStateError</code>.</p>
```
I mean we suppress rejections from *enqueued* actions past close.

adam-be replied:
This text is in line with the queued steps of addIceCandidate, but I think what you're saying makes sense. So we check for closed state right away, and then make sure that nothing is taken of the queue if the isClosed slot is true.

adam-be commented:
I've moved the rejection of closed state into the queueing steps. I've also added checks into the queuing algorithm that checks if the isClosed flag is set before advancing (e.g. taking the next operation off the queue). @jan-ivar, please review (especially the third commit).

jan-ivar commented on webrtc.html:
```diff
@@ -1300,14 +1302,14 @@
                                 <code>DOMException</code> object whose
                                 <code>name</code> attribute has the value
-                                <code>OperationError</code> and abort these
+                                <code>OperationError</code> and, abort these
```
Comma in wrong place, though I just noticed this isn't an if-statement, and there are no more steps, so can we simply remove ", and abort these steps?"

jan-ivar commented on webrtc.html:
```diff
@@ -1251,30 +1280,96 @@
+                                <p>Reject <var>p</var> with a
+                                <code>DOMException</code> object whose
+                                <code>name</code> attribute has the value
+                                <code>OperationError</code> and, abort these
```
Comma in wrong place, though I just noticed this isn't an if-statement, and there are no more steps, so can we simply remove ", and abort these steps?"

jan-ivar commented on webrtc.html:
```diff
@@ -574,6 +574,15 @@
+          <li>
+            <p>If <var>connection</var>'s [[<a>isClosed</a>]] slot is
+            <code>true</code>, return a promise rejected with an
+            <code>InvalidStateError</code>.</p>
```
I notice we're inconsistent in this spec in requiring ", and abort these steps" after `return` and `throw` statements. Do we want to clean that up separately?

jan-ivar commented on webrtc.html:
```diff
@@ -588,17 +597,37 @@
+                <code>true</code>, abort these steps.</p>
+              </li>
+              <li>
+                <p>If the promise returned by <var>operation</var> fulfilled
```
was fulfilled

jan-ivar commented on webrtc.html:
```diff
@@ -588,17 +597,37 @@
+                with a value, fulfill <var>p</var> with that value.</p>
+              </li>
+              <li>
+                <p>If the promise returned by <var>operation</var> rejected
```
was rejected

jan-ivar commented:
Lgtm with nits.

fluffy commented on webrtc.html:
```diff
@@ -1251,30 +1280,96 @@
+                                produce an identity assertion, reject
+                                <var>p</var> with a <code>DOMException</code>
+                                object whose <code>name</code> attribute has
+                                the value <code>NotReadableError</code>, and
```
It would be nice to get more information about what went wrong

adam-be replied:
Yeah. We might want to create a separate issue on that.

adam-be commented:
Thanks @jan-ivar and @fluffy for reviewing

adam-be commented:
Rebased on top of commit that patches the JSEP reference system. Also updated a JSEP reference that previously wasn't working. I think this is ready to be merged now.
